File: src/main/java/com/vectordb/TextChunker.java
package com.vectordb;

import java.util.ArrayList;
import java.util.List;

/**
 * Text chunker — splits long documents into overlapping chunks.
 * Matches the C++ chunkText() function.
 */
public final class TextChunker {

    private TextChunker() {}

    /**
     * Split text into overlapping word-based chunks.
     *
     * @param text         The full document text
     * @param chunkWords   Number of words per chunk (default: 250)
     * @param overlapWords Number of overlapping words between chunks (default: 30)
     * @return List of text chunks
     */
    public static List<String> chunkText(String text, int chunkWords, int overlapWords) {
        String[] words = text.trim().split("\\s+");
        if (words.length == 0 || (words.length == 1 && words[0].isEmpty())) {
            return List.of();
        }
        if (words.length <= chunkWords) {
            return List.of(text);
        }

        List<String> chunks = new ArrayList<>();
        int step = chunkWords - overlapWords;

        for (int i = 0; i < words.length; i += step) {
            int end = Math.min(i + chunkWords, words.length);
            StringBuilder sb = new StringBuilder();
            for (int j = i; j < end; j++) {
                if (j > i) sb.append(' ');
                sb.append(words[j]);
            }
            chunks.add(sb.toString());
            if (end == words.length) break;
        }

        return chunks;
    }

    public static List<String> chunkText(String text) {
        return chunkText(text, 250, 30);
    }
}

File: vectordb/text_chunker.py
"""
text_chunker.py — Splits long documents into overlapping word-based chunks.

Default: ~250 words per chunk, 30-word overlap between consecutive chunks.
Overlap ensures context is not lost at chunk boundaries when searching.

Equivalent to TextChunker.java in the original Java project.
"""

from typing import List


def chunk_text(
    text: str,
    chunk_words: int = 250,
    overlap_words: int = 30,
) -> List[str]:
    """
    Split a document into overlapping word-based chunks.

    Args:
        text:         The full document text.
        chunk_words:  Words per chunk (default: 250).
        overlap_words: Overlapping words between consecutive chunks (default: 30).

    Returns:
        List of text chunk strings. Always contains at least one element.
    """
    words = text.split()
    if not words:
        return [""]

    chunks: List[str] = []
    start = 0
    while start < len(words):
        end = min(start + chunk_words, len(words))
        chunk = " ".join(words[start:end])
        chunks.append(chunk)
        if end == len(words):
            break
        start += chunk_words - overlap_words  # slide window with overlap

    return chunks if chunks else [text]
